registration/api/v2/serializers.py

```diff
@@ -315,6 +344,88 @@ def validate_user(self, value):
         else:
             return value
 
+    def validate_external_domain(self, value) -> str:
+        """
+        Prevent users from registering with an external domain which was or currently in use.
+
+        The validation reduces the risk of security issues when someone is trying to take over
+        control of a client resource (domain) if they forget to restrict its access.
+        """
+        domain_data = tldextract.extract(value)
+        domain = domain_data.registered_domain
+
+        if domain == settings.DEFAULT_INSTANCE_BASE_DOMAIN:
+            raise ValidationError('The domain "{domain}" is not allowed.'.format(
+                domain=value
+            ))
+
+        self._validate_subdomain_not_contains_reserved_word(domain_data.subdomain)
+
+        is_domain_used_for_beta_app = BetaTestApplication.objects.filter(
+            external_domain__endswith=domain
+        ).exists()
+
+        if is_domain_used_for_beta_app:
+            raise ValidationError('This domain is already taken.')
+
+        is_taken = OpenEdXInstance.objects.filter(
+            Q(external_lms_domain=domain)
+            | Q(external_lms_preview_domain__endswith=domain)
+            | Q(external_studio_domain__endswith=domain)
+            | Q(external_discovery_domain__endswith=domain)
+            | Q(external_ecommerce_domain__endswith=domain)
+            # No need to check for subdomain, since it will match anyway
+            | Q(extra_custom_domains__contains=domain)
+        ).exists()
+
+        if is_taken:
+            raise ValidationError('This domain is already taken.')
+
+        return value
+
+    def validate_subdomain(self, value) -> str:
```

Comment:
@swalladge I thought about this as well. The endpoint will do the `validate_*` checks first, then (when saving the db instance) runs the model validators too.

Right now OpenCraft members would be able to create subdomains which may overwrite existing domain records, though we don't create instances from shell for customers (or testing purposes) in production environment (note that stage has a different base domain so, that wouldn't affect the DNS records).

What is your opinion @swalladge, should I merge the available subdomain validator to the existing validators? (I'm up for it since that would result in a more secure solution)

Reply:
Yep probably better to merge so all the validators are in the same place. 

> Right now OpenCraft members would be able to create subdomains which may overwrite existing domain records

I think, based on our workflows, we should restrict everyone here. This is only for the beta test application model, so if on opencraft member needs to do set a domain that wouldn't validate for any reason, they can do that through the openedx instance model.

Reply:
Consider my thoughts about validation call order invalid. I mixed up. The model validator will run first, then the serializer one.

Reply:
This is fixed in 56747bf